### PR TITLE
Upgrade to Fathom 3.0

### PR DESCRIPTION
**Pull request by biancadanforth**
Training numbers are:

**image**
```
bdanforth ~/Projects/price-tracker/src/extraction/fathom/vectors (fathom3) $ fathom-train vectors_training_image.json -a vectors_validation_image.json -l 0.07 -i 3000 -s -c "Image - early stopping"
  [#############################-------]   81%  00:03:59
Stopping early at iteration 2458, just before validation error rose.
Coeffs: [
        ['isAboveTheFoldImage', 16.586257934570312],
        ['isBig', 21.33774757385254],
        ['hasSquareAspectRatio', 1.1075100898742676],
        ['hasBackgroundInID', -10.287492752075195],
    ]
Bias: -10.176139831542969
  Training accuracy per tag:  0.99079    95% CI: (0.98879, 0.99280)  FP: 0.001  FN: 0.008
Validation accuracy per tag:  0.99335    95% CI: (0.98843, 0.99826)  FP: 0.003  FN: 0.004
  Training accuracy per page: 0.87037    95% CI: (0.81865, 0.92210)
Validation accuracy per page: 0.85000    95% CI: (0.69351, 1.00000)
```

**title**
```
bdanforth ~/Projects/price-tracker/src/extraction/fathom/vectors (fathom3) $ fathom-train vectors_training_title.json -a vectors_validation_title.json -l 0.1 -s -c "Title"
  [###############---------------------]   44%  00:00:06
Stopping early at iteration 443, just before validation error rose.
Coeffs: [
        ['isNearImageTopOrBottom', 7.078092575073242],
    ]
Bias: -1.6698582172393799
  Training accuracy per tag:  0.88934    95% CI: (0.84998, 0.92871)  FP: 0.020  FN: 0.090
Validation accuracy per tag:  1.00000    95% CI: (1.00000, 1.00000)  FP: 0.000  FN: 0.000
  Training accuracy per page: 0.98765    95% CI: (0.97065, 1.00000)
Validation accuracy per page: 1.00000    95% CI: (1.00000, 1.00000)
```

**price**
```
bdanforth ~/Projects/price-tracker/src/extraction/fathom/vectors (fathom3) $ fathom-train vectors_training_price.json -a vectors_validation_price.json -l 0.1 -c "Price" -s -v
  [####################################]  100%          
Coeffs: [
        ['hasDollarSign', 1.0177843570709229],
        ['isAboveTheFoldPrice', -5.301823616027832],
        ['hasPriceInID', 5.333859443664551],
        ['hasPriceInParentID', -7.5635271072387695],
        ['hasPriceInClassName', 1.155443787574768],
        ['hasPriceInParentClassName', 3.0024354457855225],
        ['fontIsBig', 11.338400840759277],
        ['isNearImage', 0.7539440989494324],
        ['hasPriceishPattern', 5.222956657409668],
    ]
Bias: -7.09004545211792
  Training accuracy per tag:  0.99350    95% CI: (0.99144, 0.99556)  FP: 0.002  FN: 0.004
Validation accuracy per tag:  0.99268    95% CI: (0.98685, 0.99852)  FP: 0.001  FN: 0.006
  Training accuracy per page: 0.97531    95% CI: (0.95141, 0.99921)
Validation accuracy per page: 0.95000    95% CI: (0.85448, 1.00000)
```

### CODE REVIEW

**Comment by biancadanforth** [comment]
### TL;DR:

While Price Tracker's accuracy in this PR on the 30-page test set is low at 40%, it beats out the accuracy we get on this test set in the current `master` branch which is a whopping 0%.

_Why is the test accuracy on `master` 0%?_
Well, due to a bug (see [this comment](https://github.com/mozilla/price-tracker/issues/313#issuecomment-483919114)), the `master` branch currently doesn't use Fathom at all and instead relies purely on our fallback extraction method of using CSS selectors. This PR fixes that bug, since Fathom 3.0 now scores elements based on a [0,1] confidence that the element is the target element. As a result, a `SCORE_THRESHOLD` value in `./src/extraction/fathom/index.js` was found that allows Fathom's extraction results to be used. That results in a 40% accuracy improvement!

--------------

**Test Set Results -- This PR -- Fathom accuracy: 12/30 (40%)**
The test set included 30 sample pages at random that were not a part of the training or validation sets, taken from the Fathom Commerce Samples Google Drive folder.

page | method | correct (T/F)
--- | --- | ---
0005 eBay.html | none | F
0007 Best Buy.html | fathom | T
0007 Walmart.html | fathom | T
0009 Best Buy.html | none | F
0011 eBay.html | fathom | T
0013 Best Buy.html | none | F
0014 amz.html | none | F
0015 eBay.html | none | F
00015 H_D.html | none | F
0017 Best Buy.html | none | F
amazon-40 (8).html | fathom | T
amazon-40 (22).html | none | F
amazon-40 (30).html | none | F
amazon-40 (37).html | none | F
best_buy-40 (5).html | none | F
best_buy-40 (11).html | none | F
best_buy-40 (24).html | none | F
best_buy-40 (26).html | none | F
ebay-1 (15).html | none | F
ebay-1 (30).html | none | F
home_depot-40 (17).html | none | F
home_depot-40 (21).html | fathom | T
home_depot-40 (25).html | fathom | T
home_depot-40 (28).html | fathom | T
home_depot-40 (32).html | fathom | T
home_depot-40 (38).html | none | F
walmart-40 (20).html | fathom | T
walmart-40 (25).html | fathom | T
walmart-40 (38).html | fathom | T
walmart-40 (39).html | fathom | T

----

In order to assess the accuracy of Price Tracker in this PR versus `master`, I had to run Fathom on the test set, which consisted of frozen pages saved locally (`file:///` urls). By default, Price Tracker does not perform extraction on any URLs other than `http(s):*`, so I had to apply some temporary changes to Fathom, which are documented below:

**Diff to run Fathom on `file:///*` URLs in Price Tracker**
```diff
diff --git a/src/background/index.js b/src/background/index.js
index 6861b32..17e875b 100644
--- a/src/background/index.js
+++ b/src/background/index.js
@@ -50,7 +50,7 @@ import {registerEvents, handleWidgetRemoved} from 'commerce/telemetry/extension'
   // Store the return value globally to avoid destroying it, which would
   // unregister the content scripts.
   window.registeredContentScript = browser.contentScripts.register({
-    matches: ['https://*/*', 'http://*/*'],
+    matches: ['file:///*'],
     js: [
       {file: 'extraction.bundle.js'},
     ],
@@ -60,7 +60,7 @@ import {registerEvents, handleWidgetRemoved} from 'commerce/telemetry/extension'
 
   // Set up web request listener to modify framing headers for background updates
   const webRequestFilter = {
-    urls: ['https://*/*', 'http://*/*'],
+    urls: ['file:///*'],
     types: ['sub_frame'],
     tabId: browser.tabs.TAB_ID_NONE,
   };
diff --git a/src/extraction/index.js b/src/extraction/index.js
index db15199..4f302d4 100644
--- a/src/extraction/index.js
+++ b/src/extraction/index.js
@@ -79,7 +79,9 @@ function extractProduct() {
   attempt.start();
   for (const [methodName, extract] of Object.entries(EXTRACTION_METHODS)) {
     const extractedProduct = extract(window.document);
+    console.log('methodName: ', methodName);
     if (extractedProduct) {
+      console.log('extractProduct: ', extractedProduct);
       attempt.succeed(methodName);
       return extractedProduct;
     }
@@ -117,6 +119,7 @@ async function attemptExtraction() {
 }
 
 (async function main() {
+  console.log("loaded content script to run Fathom");
   // If we're in an iframe, don't bother extracting a product EXCEPT if we were
   // started by the background script for a price check.
   const isInIframe = window !== window.top;
@@ -135,10 +138,10 @@ async function attemptExtraction() {
   // being tracked no matter what.
   const url = new URL(document.location.href);
   const allowList = await config.get('extractionAllowlist');
-  const allowAll = allowList.length === 1 && allowList[0] === '*';
-  if (!allowAll && !isBackgroundUpdate && !allowList.includes(url.host)) {
-    return;
-  }
+  // const allowAll = allowList.length === 1 && allowList[0] === '*';
+  // if (!allowAll && !isBackgroundUpdate && !allowList.includes(url.host)) {
+  //   return;
+  // }
 
   // Extract immediately, and again if the readyState changes.
   let extractedProduct = await attemptExtraction();
diff --git a/src/manifest.json b/src/manifest.json
index 4b43d27..43de2cc 100644
--- a/src/manifest.json
+++ b/src/manifest.json
@@ -27,8 +27,7 @@
     "default_popup": "browser_action/index.html"
   },
   "permissions": [
-    "https://*/*",
-    "http://*/*",
+    "file:///*",
     "tabs",
     "storage",
     "unlimitedStorage",
diff --git a/src/state/products.js b/src/state/products.js
index 840d640..1b0b52d 100644
--- a/src/state/products.js
+++ b/src/state/products.js
@@ -54,41 +54,41 @@ const ALLOWED_IMAGE_MIMETYPES = [
  * @return {Boolean}
  */
 export function isValidExtractedProduct(extractedProduct) {
-  if (typeof extractedProduct.title !== 'string') {
-    return false;
-  }
-
-  if (Number.isNaN(Date.parse(extractedProduct.date))) {
-    return false;
-  }
-
-  if (typeof extractedProduct.price !== 'number') {
-    return false;
-  }
-
-  let url;
-  let imageUrl;
-
-  try {
-    url = new URL(extractedProduct.url);
-    imageUrl = new URL(extractedProduct.image);
-  } catch (error) {
-    return false;
-  }
-
-  if (!['https:', 'http:'].includes(url.protocol)) {
-    return false;
-  }
-
-  // Some sites like Amazon initially load a data URI for the image
-  if (!['https:', 'http:', 'data:'].includes(imageUrl.protocol)) {
-    return false;
-  }
-
-  // Ensure the data URI is an image by validating its mimetype
-  if (imageUrl.protocol === 'data:' && !(ALLOWED_IMAGE_MIMETYPES.find(path => imageUrl.pathname.startsWith(path)))) {
-    return false;
-  }
+  // if (typeof extractedProduct.title !== 'string') {
+  //   return false;
+  // }
+
+  // if (Number.isNaN(Date.parse(extractedProduct.date))) {
+  //   return false;
+  // }
+
+  // if (typeof extractedProduct.price !== 'number') {
+  //   return false;
+  // }
+
+  // let url;
+  // let imageUrl;
+
+  // try {
+  //   url = new URL(extractedProduct.url);
+  //   imageUrl = new URL(extractedProduct.image);
+  // } catch (error) {
+  //   return false;
+  // }
+
+  // if (!['https:', 'http:'].includes(url.protocol)) {
+  //   return false;
+  // }
+
+  // // Some sites like Amazon initially load a data URI for the image
+  // if (!['https:', 'http:', 'data:'].includes(imageUrl.protocol)) {
+  //   return false;
+  // }
+
+  // // Ensure the data URI is an image by validating its mimetype
+  // if (imageUrl.protocol === 'data:' && !(ALLOWED_IMAGE_MIMETYPES.find(path => imageUrl.pathname.startsWith(path)))) {
+  //   return false;
+  // }
 
   return true;
 }

```

File: src/extraction/fathom/ruleset_factory.js [hunk 1]
```diff
@@ -48,18 +26,18 @@ export default class RulesetFactory {
...
   fontIsBig(fnode) {
     const size = parseInt(getComputedStyle(fnode.element).fontSize, 10);
-    return linearScale(size, 14, 50) ** this.bigFontCoeff;
+    return linearScale(size, 14, 50);
   }
 
   /** Scores fnode with a '$' in its innerText */
   hasDollarSign(fnode) {
-    return (fnode.element.innerText.includes('$') ? ONEISH : ZEROISH) ** this.hasDollarSignCoeff;
+    return (fnode.element.innerText.includes('$') ? ONEISH : ZEROISH);
```
Comment (erikrose) [inline src/extraction/fathom/ruleset_factory.js]: We should get rid of ONEISH and ZEROISH. They don't make anything better anymore, and they make make things ever so slightly worse (probably just slightly slower to converge).

File: src/extraction/fathom/ruleset_factory.js [hunk 2]
```diff
@@ -247,93 +225,86 @@ export default class RulesetFactory {
...
   hasSquareAspectRatio(fnode) {
-    return linearScale(this.aspectRatio(fnode.element), 10, 5) ** this.extremeAspectCoeff;
+    return linearScale(this.aspectRatio(fnode.element), 10, 5);
   }
 
   /**
   * Using coefficients passed into the constructor method, returns a weighted
   * ruleset used to score elements in an HTML document.
+  *
+  * @param {Array[]} An array of [string, number] tuples where the first element
```
Comment (erikrose) [inline src/extraction/fathom/ruleset_factory.js]: Is it legal to leave out the param name? In any case, we should either document both params or neither.

**Comment by biancadanforth** [comment]
**TL;DR**: Using the new [`fathom-test`](https://github.com/mozilla/fathom/pull/112) CLI on the 20 sample pages in our test set in this PR, the overall per-page accuracy of the current ruleset, weights and biases using Fathom 3.0 is 70%. This is the accuracy for Fathom to correctly identify a "product" on a page (i.e. correctly identifying all three product features: image, title and price).

**Longer version**
Using the `src/extraction/fathom/vectors/vectors_test_*` files pushed here and the coefficients and biases resulting from [training](https://github.com/mozilla/price-tracker/pull/317#issue-292567387), I used the new [`fathom-test` CLI](https://github.com/mozilla/fathom/pull/112) to measure the test accuracy (per page) for each feature against 20 sample pages:
* "image": 75%
* "title": 100%
* "price": 95%

This is not an apples to apples comparison from the past [test accuracy measure](https://github.com/mozilla/price-tracker/pull/317#issuecomment-507468069), since that was looking at whether a "product" (being the sum of its "image", "price" and "title") was found on a page. `fathom-test` checks per feature if Fathom was correct on a page or for a particular element. To directly compare, I'd need to see for which pages all 3 features were a "success" in choosing the correct element. If I look at that, the overall accuracy of Fathom in this PR to identify a product correctly (to correctly identify all three features on the page) is 70%.

This is much better than the 40% from my previous approach for testing accuracy (granted, that was on a different test set of 40 pages), but it's still nowhere near the upper 80s and 90s percent accuracy seen in the training and validation runs, despite these samples coming from the exact same corpus as the training and validation samples.

**Image testing accuracy**
```bash
(venv) bdanforth ~/Projects/price-tracker (fathom3) $ fathom-test src/extraction/fathom/vectors/vectors_test_image.json '{"coeffs": [["isAboveTheFoldImage", 16.586257934570312], ["isBig", 21.33774757385254], ["hasSquareAspectRatio", 1.1075100898742676], ["hasBackgroundInID", -10.287492752075195]], "bias": -10.176139831542969}'

Testing accuracy per tag:  0.99000    95% CI: (0.98412, 0.99588)  FP: 0.003  FN: 0.007
Testing accuracy per page: 0.75000    95% CI: (0.56022, 0.93978)

Testing per-page results:
 success  on     amazon-40%20(22).html. Confidence: 0.45426598
 success  on     amazon-40%20(30).html. Confidence: 0.79204732
 failure  on     amazon-40%20(37).html. Confidence: 0.71261239 Highest-scoring element was a wrong choice.
    First target at index 7: 0.18024161
 success  on      amazon-40%20(8).html. Confidence: 0.45773035
 failure  on   best_buy-40%20(11).html. Confidence: 0.06214701 Highest-scoring element was a wrong choice.
    First target at index 2: 0.03152284
 success  on   best_buy-40%20(24).html. Confidence: 0.31307796
 failure  on   best_buy-40%20(26).html. Confidence: 0.06214701 Highest-scoring element was a wrong choice.
    First target at index 2: 0.05037770
 success  on    best_buy-40%20(5).html. Confidence: 0.10311700
 success  on        ebay-1%20(15).html. Confidence: 0.20423122
 success  on        ebay-1%20(30).html. Confidence: 0.81114835
 failure  on home_depot-40%20(17).html. Confidence: 0.74345410 Highest-scoring element was a wrong choice.
    First target at index 1: 0.73110878
 success  on home_depot-40%20(21).html. Confidence: 0.73110878
 success  on home_depot-40%20(25).html. Confidence: 0.97135264
 success  on home_depot-40%20(28).html. Confidence: 0.73110878
 success  on home_depot-40%20(32).html. Confidence: 0.73110878
 failure  on home_depot-40%20(38).html. Confidence: 0.74345410 Highest-scoring element was a wrong choice.
    First target at index 1: 0.73110878
 success  on    walmart-40%20(20).html. Confidence: 0.60515302
 success  on    walmart-40%20(25).html. Confidence: 0.60668099
 success  on    walmart-40%20(38).html. Confidence: 0.58588487
 success  on    walmart-40%20(39).html. Confidence: 0.53866470
```

**Title testing accuracy**
```bash
(venv) bdanforth ~/Projects/price-tracker (fathom3) $ fathom-test src/extraction/fathom/vectors/vectors_test_title.json '{"coeffs": [["isNearImageTopOrBottom", 7.078092575073242]], "bias": -1.6698582172393799}'

Testing accuracy per tag:  0.83871    95% CI: (0.70923, 0.96818)  FP: 0.000  FN: 0.161
Testing accuracy per page: 1.00000    95% CI: (1.00000, 1.00000)

Testing per-page results:
 success  on     amazon-40%20(22).html. Confidence: 0.99099052
 success  on     amazon-40%20(30).html. Confidence: 0.64029801
 success  on     amazon-40%20(37).html. Confidence: 0.24906397
 success  on      amazon-40%20(8).html. Confidence: 0.97334236
 success  on   best_buy-40%20(11).html. Confidence: 0.24906397
 success  on   best_buy-40%20(24).html. Confidence: 0.69247615
 success  on   best_buy-40%20(26).html. Confidence: 0.24906397
 success  on    best_buy-40%20(5).html. Confidence: 0.76132601
 success  on        ebay-1%20(15).html. Confidence: 0.99085999
 success  on        ebay-1%20(30).html. Confidence: 0.99085999
 success  on home_depot-40%20(17).html. Confidence: 0.24906397
 success  on home_depot-40%20(21).html. Confidence: 0.98728508
 success  on home_depot-40%20(25).html. Confidence: 0.82306099
 success  on home_depot-40%20(28).html. Confidence: 0.93691045
 success  on home_depot-40%20(32).html. Confidence: 0.93691045
 success  on home_depot-40%20(38).html. Confidence: 0.24906397
 success  on    walmart-40%20(20).html. Confidence: 0.98574233
 success  on    walmart-40%20(25).html. Confidence: 0.97179431
 success  on    walmart-40%20(38).html. Confidence: 0.98574233
 success  on    walmart-40%20(39).html. Confidence: 0.98574233
```

**Price testing accuracy**
```bash
(venv) bdanforth ~/Projects/price-tracker (fathom3) $ fathom-test src/extraction/fathom/vectors/vectors_test_price.json '{"coeffs": [["hasDollarSign", 1.0177843570709229], ["isAboveTheFoldPrice", -5.301823616027832], ["hasPriceInID", 5.333859443664551], ["hasPriceInParentID", -7.5635271072387695], ["hasPriceInClassName", 1.155443787574768], ["hasPriceInParentClassName", 3.0024354457855225], ["fontIsBig", 11.338400840759277], ["isNearImage", 0.7539440989494324], ["hasPriceishPattern", 5.222956657409668]], "bias": -7.09004545211792}'

Testing accuracy per tag:  0.99458    95% CI: (0.98847, 1.00000)  FP: 0.000  FN: 0.005
Testing accuracy per page: 0.95000    95% CI: (0.85448, 1.00000)

Testing per-page results:
 failure  on     amazon-40%20(22).html. Confidence: 0.01885081 Highest-scoring element was a wrong choice.
    First target at index 1: 0.01707094
 success  on     amazon-40%20(30).html. Confidence: 0.01656970 No target nodes. Assumed negative sample.
 success  on     amazon-40%20(37).html. Confidence: 0.01240653 No target nodes. Assumed negative sample.
 success  on      amazon-40%20(8).html. Confidence: 0.82405251
 success  on   best_buy-40%20(11).html. Confidence: no candidate nodes. Assumed negative sample.
 success  on   best_buy-40%20(24).html. Confidence: 0.88195562
 success  on   best_buy-40%20(26).html. Confidence: no candidate nodes. Assumed negative sample.
 success  on    best_buy-40%20(5).html. Confidence: 0.88763416
 success  on        ebay-1%20(15).html. Confidence: 0.15543531 No target nodes. Assumed negative sample.
 success  on        ebay-1%20(30).html. Confidence: 0.14137758
 success  on home_depot-40%20(17).html. Confidence: no candidate nodes. Assumed negative sample.
 success  on home_depot-40%20(21).html. Confidence: 0.99791569
 success  on home_depot-40%20(25).html. Confidence: 0.99840385
 success  on home_depot-40%20(28).html. Confidence: 0.99801970
 success  on home_depot-40%20(32).html. Confidence: 0.99818760
 success  on home_depot-40%20(38).html. Confidence: no candidate nodes. Assumed negative sample.
 success  on    walmart-40%20(20).html. Confidence: 0.82440346
 success  on    walmart-40%20(25).html. Confidence: 0.79404730
 success  on    walmart-40%20(38).html. Confidence: 0.81192786
 success  on    walmart-40%20(39).html. Confidence: 0.83670110
```

**Comment by erikrose** [comment]
> This is the accuracy for Fathom to correctly identify a "product" on a page (i.e. correctly identifying all three product features: image, title and price).

How did you measure overall "product" success? Did you do manual math to intersect the per-page successes of the 3 types?

> This is not an apples to apples comparison from the past test accuracy measure, since that was looking at whether a "product" (being the sum of its "image", "price" and "title") was found on a page.

I suspect from your linked description that you manually kept track of successes on the hacked-up copy of Price Tracker to determine the past score. Is that true? (I want to make sure you understand the old FathomFox Trainer also tested one type at a time.)

**Comment by biancadanforth** [comment]
> How did you measure overall "product" success? Did you do manual math to intersect the per-page successes of the 3 types?

Yes. That may be a nice enhancement for the `fathom-web` CLI to have an option to do this for you in applications where we need multiple features to be correct on a page to consider it a success.

> I suspect from your linked description that you manually kept track of successes on the hacked-up copy of Price Tracker to determine the past score. Is that true? (I want to make sure you understand the old FathomFox Trainer also tested one type at a time.)

I did manually keep track of successes, but I'm not sure what you mean by "hacked up copy" -- the copy of Price Tracker I used was this PR.

I do understand that Fathom's current and previous testing method runs on a per-feature basis. Sorry, I misspoke a bit when I said:
> ... but it's still nowhere near the upper 80s and 90s percent accuracy seen in the training and validation runs, despite these samples coming from the exact same corpus as the training and validation samples.

What I meant to say was that, while "title" and "price" testing accuracy (per page) is within a few percentage points of their training and validation accuracy, "image" testing accuracy per page is a full 10% lower at 75%.

**Comment by erikrose** [comment]
"Hacked up": I was referring to the temporary changes you made in https://github.com/mozilla/price-tracker/pull/317#issuecomment-507468069. I also had in mind some more extensive changes you'd made, but that must have been in a similar ticket. So it's not that hacked-up after all. :-)

> What I meant to say was that, while "title" and "price" testing accuracy (per page) is within a few percentage points of their training and validation accuracy, "image" testing accuracy per page is a full 10% lower at 75%.

Oh, good. Now we're getting into the realm of explicability. Could be legit unluckiness at this point. For a 10% change, we'd have only to do worse on 2 samples out of the 20 used.
